src/exporter/DaemonMetricCollector.cc

```diff
@@ -0,0 +1,201 @@
+void add_metric(std::stringstream &ss, T value, std::string name,
+                std::string description, std::string mtype,
+                std::string labels) {
+  ss << "# HELP " << name << " " << description << "\n";
+  ss << "# TYPE " << name << " " << mtype << "\n";
+  ss << name << "{" << labels << "} " << value << "\n";
+}
+
+void add_double_or_int_metric(std::stringstream &ss, json_value value,
+                              std::string name, std::string description,
+                              std::string mtype, std::string labels) {
+  if (value.is_int64()) {
+    int64_t v = value.as_int64();
+    add_metric(ss, v, name, description, mtype, labels);
+  } else if (value.is_double()) {
+    double v = value.as_double();
+    add_metric(ss, v, name, description, mtype, labels);
+  }
+}
+
+std::string boost_string_to_std(boost::json::string js) {
+  std::string res(js.data());
+  return res;
+}
+
+std::string quote(std::string value) { return "\"" + value + "\""; }
+
+bool is_hyphen(char ch) { return ch == '-'; }
+
+void DaemonMetricCollector::dump_asok_metrics() {
+  std::stringstream ss;
+  for (auto client : clients) {
+    AdminSocketClient &sock_client = client.second;
+    std::string daemon_name = client.first;
+    std::string perf_dump_response = asok_request(sock_client, "perf dump");
+    if (perf_dump_response.size() > 0) {
+      json_object dump = boost::json::parse(perf_dump_response).as_object();
+      std::string perf_schema_response =
+          asok_request(sock_client, "perf schema");
+      json_object schema = boost::json::parse(perf_schema_response).as_object();
+      for (auto perf : schema) {
+        std::string perf_group = perf.key().to_string();
+        json_object perf_group_object = perf.value().as_object();
+        for (auto perf_counter : perf_group_object) {
+          std::string perf_name = perf_counter.key().to_string();
+          json_object perf_info = perf_counter.value().as_object();
+          if (perf_info["priority"].as_int64() <
+              PerfCountersBuilder::PRIO_USEFUL) {
+            continue;
+          }
+          std::string name = "ceph_" + perf_group + "_" + perf_name;
+          std::replace_if(name.begin(), name.end(), is_hyphen, '_');
+
+          // FIXME: test this, based on mgr_module perfpath_to_path_labels
+          auto labels_and_name = get_labels_and_metric_name(daemon_name, name);
+          std::string labels = labels_and_name.first;
+          name = labels_and_name.second;
+
+          json_value perf_values = dump[perf_group].as_object()[perf_name];
+          dump_asok_metric(ss, perf_info, perf_values, name, labels);
+        }
+      }
+    }
+  }
+  metrics = ss.str();
+}
+
+std::string DaemonMetricCollector::asok_request(AdminSocketClient &asok,
+                                                std::string command) {
+  std::string request("{\"prefix\": \"" + command + "\"}");
+  std::string response;
+  asok.do_request(request, &response);
```

Comment:
we need to handle errors from `do_request()` here. we should expect admin sockets to go away randomly as daemons are started/stopped

on errors, it looks like `do_request()` will return a non-empty std::string with the error message